storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +64,395 @@
-            }
+    private void markInitialized(final UserTopicIdPartition utp) {
+        // Silently not initialize the utp
+        if (!utp.isAssigned) {
+            log.warn("Tried to initialize a UTP: {} that was not yet assigned!", utp);
+            return;
+        }
+        if (!utp.isInitialized) {
+            remotePartitionMetadataEventHandler.markInitialized(utp.topicIdPartition);
+            utp.isInitialized = true;
         }
     }
 
-    public Set<Integer> metadataPartitionsAssigned() {
-        return Collections.unmodifiableSet(assignedMetaPartitions);
+    static Set<TopicPartition> toRemoteLogPartitions(final Set<Integer> partitions) {
+        return partitions.stream()
+            .map(ConsumerTask::toRemoteLogPartition)
+            .collect(Collectors.toSet());
+    }
+
+    static TopicPartition toRemoteLogPartition(int partition) {
+        return new TopicPartition(REMOTE_LOG_METADATA_TOPIC_NAME, partition);
+    }
+
+    static class UserTopicIdPartition {
+        private final TopicIdPartition topicIdPartition;
+        private final Integer metadataPartition;
+        // The `utp` will be initialized once it reads all the existing events from the remote log metadata topic.
+        boolean isInitialized;
+        // denotes whether this `utp` is assigned to the consumer
+        boolean isAssigned;
+
+        /**
+         * UserTopicIdPartition denotes the user topic-partitions for which this broker acts as a leader/follower of.
+         *
+         * @param tpId               the unique topic partition identifier
+         * @param metadataPartition  the remote log metadata partition mapped for this user-topic-partition.
+         */
+        public UserTopicIdPartition(final TopicIdPartition tpId, final Integer metadataPartition) {
+            this.topicIdPartition = Objects.requireNonNull(tpId);
+            this.metadataPartition = Objects.requireNonNull(metadataPartition);
+            this.isInitialized = false;
+            this.isAssigned = false;
+        }
+
+        @Override
+        public String toString() {
+            return "UserTopicIdPartition{" +
+                "topicIdPartition=" + topicIdPartition +
+                ", metadataPartition=" + metadataPartition +
+                ", isInitialized=" + isInitialized +
+                ", isAssigned=" + isAssigned +
+                '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            UserTopicIdPartition that = (UserTopicIdPartition) o;
+            return topicIdPartition.equals(that.topicIdPartition) && metadataPartition.equals(that.metadataPartition);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(topicIdPartition, metadataPartition);
+        }
+    }
+
+    static class StartAndEndOffsetHolder {
```

Comment:
nit: not sure the `Holder` adds much, but feel free to ignore this.
```suggestion
    static class StartAndEndOffsets {
```